library/std/src/sys/process/unix/unix.rs

```diff
@@ -397,9 +402,69 @@ impl Command {
 
             _reset = Some(Reset(*sys::env::environ()));
             *sys::env::environ() = envp.as_ptr();
+            // We've set the environment, no need to set it again.
+            maybe_envp = None;
         }
 
-        libc::execvp(self.get_program_cstr().as_ptr(), self.get_argv().as_ptr());
+        use crate::ffi::CStr;
+        // execvpe is a gnu extension...
+        #[cfg(all(target_os = "linux", target_env = "gnu"))]
+        unsafe fn exec_with_env(
+            program: &CStr,
+            args: &CStringArray,
+            envp: &CStringArray,
+        ) -> libc::c_int {
+            unsafe { libc::execvpe(program.as_ptr(), args.as_ptr(), envp.as_ptr()) }
+        }
+
+        // ...so if we're not gnu then use our own implementation.
+        #[cfg(not(all(target_os = "linux", target_env = "gnu")))]
+        unsafe fn exec_with_env(
+            program: &CStr,
+            args: &CStringArray,
+            envp: &CStringArray,
+        ) -> libc::c_int {
+            unsafe {
+                let name = program.to_bytes();
+                let mut buffer =
+                    [const { mem::MaybeUninit::<u8>::uninit() }; libc::PATH_MAX as usize];
```

Comment:
I'm aware that paths can be longer than `PATH_MAX` but I can't allocate here. I could allocate a buffer before fork but this is an experimental feature so I really want it to be as self-contained as possible so it can be removed or radically changed without too much disruption. ~~Unsized locals would be another alternative...~~

In any case, if this feature ever approaches stabilisation, I can fix it then.